4단계 - 볼링 점수판(n명)

jhwlim opened the pull request:
안녕하세요. 4단계 리뷰 요청드립니다.

3단계 리뷰에서 마지막으로 남기신 코멘트 관련 답변드립니다.
1. https://github.com/next-step/java-bowling/pull/966#discussion_r1022755885
    - 투구 상태는 1 ~ 9 프레임에서는 첫번째 투구시 쓰러트린 핀의 개수가 1 ~ 9개라면, null 이 나올 수 있고, 10 프레임에서도 첫번째 투구 또는 핀을 모두 다 쓰러트린 다음 투구시 1 ~ 9개의 핀을 쓰러트리면 null 이 나올 수 있어서 별도로 수정하지는 않았습니다. 다만, null 을 사용하는 것은 지양하는게 좋을 것 같아서, NONE과 같은 상태 클래스를 별도로 만들어야 할지 고민을 하기는 했는데, 일단은 그냥 null 을 사용했습니다.
2. https://github.com/next-step/java-bowling/pull/966#discussion_r1022765458
    - HitState 내부에서 hits 를 관리하는 것을 생각해봤는데, HitState에 hit가 있는 것 보다, Hit에 HitState가 있는 것이 맞지 않을까해서 각 투구를 관리하는 Hit 클래스를 추가했고, 투구와 관련된 밸리데이션은 Hit 클래스 내부로 이동했습니다.

jhwlim committed:
docs : 기능 목록 정리

jhwlim committed:
feat : 여러 명의 사용자가 볼링게임 점수판을 가질 수 있도록 수정

jhwlim committed:
test : Player 테스트 추가

jhwlim committed:
refactor : 플레이어 입력 메서드 이동

jhwlim committed:
test : HitState 테스트 추가

jhwlim committed:
refactor : 프로덕션 환경에서 사용하지 않는 메서드 삭제

jhwlim committed:
refactor : Hit 클래스 분리

jhwlim committed:
feat : 볼링 게임 참여자 숫자 0 보다 큰 숫자만 입력하도록 수정

seokhongkim commented on src/main/java/bowling/views/InputView.java:
```diff
@@ -9,13 +13,30 @@ public class InputView {
-    public static String inputPlayerName() {
+    public static Players inputPlayers() {
+        int numberOfPlayers = InputView.inputNumberOfPlayers();
+        Players players = new Players();
```
`Players` 객체는 `players.add` 를 통해 플레이어가 추가된 이후에만 유의미하게 사용될 수 있을 것 같아요.
게임에 참가할 참여자 정보를 모두 입력받은 후 생성자에서 `players` 목록을 전달받으면 어떨까요? 이렇게 하면 생성된 이후부터 유의미한 값을 가질 수 있고, 이후에는 불변 상태를 유지할 수 있을 것 같네요!

seokhongkim commented on src/main/java/bowling/BowlingGameApplication.java:
```diff
@@ -18,15 +22,34 @@ public static void main(String[] args) {
         }
     }
 
+    private static void playFrame(Players players, int index) {
```
`playPlayerTurn` 과 같이 메서드 이름을 `playFrame` 와 다르게 가져가면 어떨까요?
다른 기능을 하는데 동일한 이름이라서 마치 오버로딩한 것처럼 보이네요!

seokhongkim commented on src/main/java/bowling/domain/Player.java:
```diff
@@ -7,10 +7,12 @@ public class Player {
     private static final String NAME_REGEX = "^[A-Za-z]{3}$";
 
     private final String name;
+    private final BowlingGame game;
```
플레이어가 볼링 게임을 관리하도록 변경하여 플레이어만 늘렸을 때 N명이 플레이할 수 있는 구조를 만드셨군요 👍

seokhongkim commented on src/test/java/bowling/domain/state/HitStateTest.java:
```diff
@@ -15,6 +16,19 @@ void from_givenStrike() {
         assertThat(HitState.from(10, null)).isEqualTo(HitState.STRIKE);
     }
 
+    @DisplayName("첫번째 시도에서 1개 이상, 10개 미만의 핀을 쓰러트리지 못하면, null을 반환해야 한다.")
```
`쓰러트리지 못하면,` -> `쓰러트리면,` 이겠죠?

jhwlim replied:
앗 넵 맞습니다! 수정하겠습니다;;

seokhongkim commented on src/main/java/bowling/domain/Hit.java:
```diff
@@ -0,0 +1,83 @@
+        }
+
+        int hit = previousHit.value;
+        if (hit == Frame.MAX_NUMBER_OF_BOWLING_PINS) {
```
`hit` 이 `MAX_NUMBER_OF_BOWLING_PINS` 와 같은데, `STRIKE` 또는 `SPARE` 가 아닌 케이스가 발생할 수 있나요? :)

jhwlim replied:
아 생각해보니 그렇네요... state 따로 확인하고 value로 따로 확인하다보니 중복이 있었던 것 같습니다. 수정하겠습니다!

seokhongkim commented on src/main/java/bowling/domain/Players.java:
```diff
@@ -0,0 +1,26 @@
+import java.util.ArrayList;
+import java.util.List;
+
+public class Players {
```
현재는 `List` 대신 일급 컬렉션을 사용해서 얻을 수 있는 이점이 없는 것 같네요.
플레이어의 순서를 관리한다거나, 이름을 통해 플레이어를 가져오는 등의 역할을 가져볼 수도 있을 것 같습니다. `index` 로 직접 접근한다면 `List` 인 자료구조를 그대로 사용한 것과 차이가 없는 것 같아요. :)

seokhongkim commented on src/main/java/bowling/domain/frame/FinalFrame.java:
```diff
@@ -5,14 +5,19 @@
+        if (hits.size() < MIN_SIZE_OF_HITS) {
+            return false;
+        }
+        return HitState.MISS.equals(getState(MIN_SIZE_OF_HITS - 1))
```
`MIN_SIZE_OF_HITS` 가 2라는 값이라서 여기에서 사용이 된 것일까요? 두 번째 hit 의 상태가 `MISS` 인 경우 종료를 나타내기 위한 것일까요? 그렇다면 명시적으로 두 번째 투구의 상태를 가져와서 비교해보면 어떨까요?

jhwlim replied:
두번째 hit의 상태가 MISS인 경우 종료를 나타내기 위한 조건으로 매직넘버를 없애려고 위와 같이 작성했는데, 다시보니 의미 전달이 불분명한 것 같습니다;; 수정하겠습니다!

seokhongkim reviewed (changes requested):
종환님, 4단계 미션 잘 구현해주셨습니다. 👍 

코드를 보고 궁금한 점과 고민해보시면 좋을 것 같은 부분을 코멘트로 남겼습니다.
코멘트 확인 부탁드립니다. 🙏

jhwlim committed:
refactor : 피드백 반영

- Players 클래스 삭제
- Hit 내부의 남은 핀의 개수를 구하는 로직 중 중복으로 체크하는 부분 수정
- 메서드 이름 명확하게 수정
- 오타 수정

jhwlim commented:
이번 단계는 디테일한 부분을 많이 놓친 것 같은데, 피드백 감사드립니다!
피드백 주신 내용 참고하여 내용 수정했는데, 혹시 추가로 수정할 만한 부분 있으면 다시 코멘트 부탁드립니다!

seokhongkim commented on src/main/java/bowling/BowlingGameApplication.java:
```diff
@@ -22,21 +20,21 @@ public static void main(String[] args) {
     }
 
-    private static void playFrame(Players players) {
+    private static void playFrameTurn(List<Player> players) {
```
IDE가 메서드 리네임 과정에서 오버로딩한 메서드로 인식해 두 메서드의 이름을 함께 변경한 것 같네요!

seokhongkim reviewed (approved):
종환님, 4단계 미션 코멘트도 잘 반영해주셨습니다. 💯
4단계 PR도 이만 머지하도록 하겠습니다.
코멘트 반영해주신 부분 중 의도와 다르게 수정된 코드로 보이는 곳이 있어 코멘트 하나 남겼습니다. 참고 부탁드립니다. 🙏 

교육 과정의 마지막 미션까지 잘 마무리하셨습니다.
볼링 점수판 미션을 진행하며 클래스와 메서드의 네이밍, 클래스의 책임, 의존 관계, 값 객체 등을 중심으로 코멘트를 드렸는데, 잘 전달이 되어 학습하시는 데 도움이 되었으면 하는 바람입니다. :)
미션 진행하는 동안 수고 많으셨습니다! 👏
